docs: add SSR page (#473)
tuono-labs/tuono: apps/documentation/src/routes/documentation/rendering/server-side-rendering.mdx

<Breadcrumbs breadcrumbs={[{ label: 'Rendering' }]} />

# Server side rendering

## Overview

If a page uses **Server-side Rendering**, the page HTML is generated on **each request**.

> > Every page is automatically server side rendered by default.

Reviewer: ```suggestion
> By default, every page is automatically server side rendered.
```

> To enable SSR on a page you just need to create the same file with the rust extension:

Reviewer 2: Just a quick one - on line 18, you say by default, but here you say it needs the Rust file. Which one? Great doc though :)

Author: You right! Better being more specific! Thanks for checking

- `src/routes/about.tsx`
- `src/routes/about.rs`

> and then create an `#[tuono_lib::handler]` in the rust file that will be the
> function actually called on each request before rendering the HTML.

Reviewer: ```suggestion
and then create a #[tuono_lib::handler] in the rust file, 
which will be the function called on each request before rendering the HTML.
```

## Example

> For example, suppose that your page needs to pre-render frequently updated data
> (read from a local file). You can write an hanlder which reads a json file and passes
> it to the route like below:

Reviewer: ```suggestion
For example, suppose your page needs to pre-render frequently updated data
(read from a local file). 
You can write a handler which reads a JSON file and passes it to the route like below:
```

```rs
// src/routes/about.rs
use tuono_lib::{Request, Response, Props};
use std::fs;

#[tuono_lib::handler]
async fn my_custom_ssr_page(req: Request) -> Response {
  let data = fs::read_to_string("my_data.json").expect("Failed to read json file");

>   Response::Props(Props::new(data))

Reviewer: I haven't used this feature so far but I have some doubts about `Props` name usage:
I don't know how much familiar this term can be for user when dealing with http responses and with framework different from React.
Since you plan to make `tuono` framework agnostic I would consider another name.
I would use something tied to HTTP response statuses like
`OK` (200) since another option for Response is `Redirect`.
If this isn't ok here are a few additional alternatives:
- `Content`
- `Output`
- `RouteData`

(We can address this in a different PR)

Author: Agree. I'd propose to do an investigation about the overall DX and syntax! Agree that it is pretty rough right now

}
```

Then you will be able to read the loaded data on the route in the route's `data` props:

```tsx
// src/routes/about.tsx

> import type { TuonoProps } from 'tuono'

Reviewer: I already bring this up in #450, but IMHO we should give a more semantic name to this type.

(We can address this in a different PR)

export default function AbouteRoute({ data }: TuonoProps) {
  // ...
}
```

> > `TuonoProps<T>` takes a generic argument for statically typing `data`.

Reviewer: ```suggestion
> `TuonoProps<T>` takes a generic argument to statically type `data`.
```
